angr/sim_manager.py

```diff
@@ -0,0 +1,927 @@
+        :param state_ranker:    An alternative to stash_splitter. States will be sorted with outputs
+                                of this function, which are to be used as a key. The first "limit"
+                                of them will be kept, the rest split off.
+        :param limit:           For use with state_ranker. The number of states to keep. Default: 8
+        :param from_stash:      The stash to split (default: 'active')
+        :param to_stash:        The stash to write to (default: 'stashed')
+
+        :returns:               The resulting SimulationManager.
+        :rtype:                 SimulationManager
+        """
+        from_stash = from_stash or 'active'
+        to_stash = to_stash or 'stashed'
+        states = self._fetch_states(stash=from_stash)
+
+        if stash_splitter is not None:
+            keep, split = stash_splitter(states)
+        elif stash_ranker is not None:
+            ranked_paths = stash_ranker(states)
+            keep, split = ranked_paths[:limit], ranked_paths[limit:]
+        elif state_ranker is not None:
+            ranked_paths = sorted(states, key=state_ranker)
+            keep, split = ranked_paths[:limit], ranked_paths[limit:]
+        else:
+            keep, split = states[:limit], states[limit:]
+
+        keep, split = map(list, (keep, split))
+
+        self._clear_states(from_stash)
+        self._store_states(from_stash, keep)
+        self._store_states(to_stash, split)
+        return self
+
+    @ImmutabilityMixin.immutable
+    def merge(self, merge_func=None, stash=None):
+        """Merge the states in a given stash.
+
+        :param stash:       The stash (default: 'active')
+        :param merge_func:  If provided, instead of using state.merge, call this function with
+                            the states as the argument. Should return the merged state.
+
+        :returns:           The result SimulationManager.
+        :rtype:             SimulationManager
+        """
+        stash = 'active' if stash is None else stash
+        self.prune(from_stash=stash)
+        to_merge = self._fetch_states(stash=stash)
+        not_to_merge = []
+
+        merge_groups = [ ]
+        while to_merge:
+            g, to_merge = self._filter_states(lambda s: s.addr == to_merge[0].addr, to_merge)
+            if len(g) <= 1:
+                not_to_merge.extend(g)
+            else:
+                merge_groups.append(g)
+
+        for g in merge_groups:
+            try:
+                m = self._merge_states(g) if merge_func is None else merge_func(*g)
+                not_to_merge.append(m)
+            except SimMergeError:
+                l.warning("SimMergeError while merging %d states", len(g), exc_info=True)
+                not_to_merge.extend(g)
+
+        self._clear_states(stash)
+        self._store_states(stash, not_to_merge)
+        return self
+
+    #
+    # Various canned functionality (TBH, I don't quite see, why these exists.)
```

Comment:
the why is that we weren't quite sure how simgr would be used when we were first writing it, and we tossed in this functionality because it was easy to do and we thought it would be useful. we were wrong, so far as we can tell!

Reply:
So, I think, this "coffee-making" methods should be removed as well?

Reply:
Yeah, we should remove them. When I wrote PathGroups, I sat down and brainstormed ways one would use them, and then for some reason, implemented all of them as helper functions, but I don't think these functions were ever very useful.